## Упражнения по библиотеке Numpy

In [2]:
import numpy as np
import numpy.random as rand

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [30]:
def random_array():
    return np.random.randint(-10, 10, size=10)

In [31]:
rand_arr = random_array()
print(rand_arr)
rand_arr = np.where(np.logical_and(rand_arr>3,rand_arr<8), rand_arr*-1, rand_arr)
print(rand_arr)

[ -5   5 -10   0   6   7   6  -1  -6 -10]
[ -5  -5 -10   0  -6  -7  -6  -1  -6 -10]


**2.** Заменить максимальный элемент случайного массива на 0

In [40]:
a = random_array()
print(a)
a = np.where(a == max(a), 0, a)
print(a)

[ 5 -6 -2 -5  7  2  3 -7  7 -2]
[ 5 -6 -2 -5  0  2  3 -7  0 -2]


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [67]:
def cart(a):
    arrays = [np.asarray(i) for i in a]
    matrix = map(len, arrays)
    indexes = np.indices(matrix, dtype=int)
    true_indexes = indexes.reshape(len(arrays), -1).T
    for n, arr in enumerate(arrays):
        true_indexes[:, n] = arrays[n][true_indexes[:, n]]
    return true_indexes


cart([[1, 2, 3], [4, 5, 6]])

array([[1, 4],
       [1, 5],
       [1, 6],
       [2, 4],
       [2, 5],
       [2, 6],
       [3, 4],
       [3, 5],
       [3, 6]])

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [88]:
A = np.random.randint(-10,10,(8,3))
B = np.random.randint(-10,10,(2,2))
nn = (A[..., np.newaxis, np.newaxis] == B)
res = (nn.sum(axis=(1,2,3))  >= B.shape[1]).nonzero()[0]
print(res)

[4]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [96]:
matrix = np.random.randint(-10, 10, (10,3))
equal = np.logical_and.reduce(matrix[:,1:] == matrix[:,:-1], axis=1)
res = matrix[~equal]

**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [98]:
def Unique(array):
    return np.unique(array, axis = 0)
Unique([[1,1,1],[1,1,1],[1,2,3]])

array([[1, 1, 1],
       [1, 2, 3]])

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [112]:
#!numpy == nn
res_nn = 1
X_nn = [[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]
is_exist = 0
for i in range(len(X_nn)):
    for j in range(len(X_nn[1])):
        if(i == j & X_nn[i][j] != 0):
            res_nn *= X_nn[i][j]
            is_exist = 1
if(is_exist):
    print("No numpy result is: ", res_nn)
# numpy == n
X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
res = np.prod(np.diag(X)[np.diag(X) != 0])
print("Numpy result is: ", res)

No numpy result is:  3
Numpy result is:  3


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [114]:
# nn
x = [1, 2, 2, 4]
y = [4, 2, 1, 2]
print(set(x) == set(y))
# n
x_ = np.array([1, 2, 2, 4])
y_ = np.array([4, 2, 1, 2])
print(np.array_equal(x_.sort(), y_.sort()))

True
True


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [122]:
#nn
x = [6, 2, 0, 3, 0, 0, 5, 7, 0]
l = []
for i in range(len(x) - 1):
    if (x[i] == 0 and x[i+1] != 0):
        l.append(x[i+1])
print(max(l))
#n
x_ = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
is_zero = x_ == 0
print(x_[1:][is_zero[:-1]].max())


5
5


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [151]:
#nn
x = [2, 2, 2, 3, 3, 3, 5]
set_x = set(x)
res_nn = [0 for i in set_x]
c = 0
for i in set_x:
    for j in range(len(x)):
        if(x[j] == i):
            res_nn[c] += 1
    c+=1
print(set_x,res_nn)
#n
x_ = np.array([2, 2, 2, 3, 3, 3, 5]) 
print((np.unique(x_), np.bincount(x_)[np.bincount(x_) != 0]))

{2, 3, 5} [3, 3, 1]
(array([2, 3, 5]), array([3, 3, 1], dtype=int64))


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [160]:
#Объекты одномерные
#nn
def no_numpy_cdist(X, Y):
    res_matrix = []
    res_raw = []
    for i in range(len(X)):
        for j in range(len(Y)):
              res_raw.append(abs(X[i] - Y[j]))
        res_matrix.append(res_raw)
        res_raw = []
    return res_matrix
#n
def numpy_cdist(X,Y):
    return [[abs(x - y) for y in Y] for x in X]

_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [77]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [161]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [164]:
my_calories = 60.
average_calories = np.mean(calorie_stats)
delta_calories = abs(my_calories - average_calories)
print(average_calories)

106.88311688311688


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [167]:
calorie_stats_sorted = np.sort(calorie_stats)
print(calorie_stats_sorted)

[ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [168]:
median_calories = np.median(calorie_stats)
print(median_calories)

110.0


5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [196]:
print(np.percentile(calorie_stats_sorted, 50))
print(np.percentile(calorie_stats_sorted, 25))
print(np.percentile(calorie_stats_sorted, 12.5))
print(np.percentile(calorie_stats_sorted, 6.75))
print(np.percentile(calorie_stats_sorted, 3.375))
print(np.percentile(calorie_stats_sorted, 3.2894737))
nth_percentile = 3.2894737

110.0
100.0
90.0
81.30000000000001
61.300000000000004
60.00000024


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [206]:
more_calories = (calorie_stats[calorie_stats > 60].size / calorie_stats.size) * 100
print(more_calories)

96.1038961038961


7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [208]:
calorie_std = np.std(calorie_stats_sorted)
print(calorie_std)

19.35718533390827


8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

# Вывод
## Калорий меньше, чем у 96% конкурентов
Мы нашли минимальный процентиль, превышающий 60 калорий, а также подсчитали процент хлопьев, в которых содержится более 60 калорий на порцию.
## Калорий меньше, чем величина стандартного отклонения от среднего значения
Для математиков, которые будут смотреть рекламу, станет хорошим подспорьем тот факт, что количество калорий в этом продукте меньше, чем среднее значение минус стандатрное отклонение, более широких маркетинговых бенефитов пока не вижу:)
